Dunitz/1850 genesets api (#98)
chanzuckerberg/single-cell: rfcs/0003-gene-sets/api.md

are selected and the difference in the expression level for each gene is computed by comparing the average
expression level for each set of cells. Here I will define the structure of the API to ensure communication between the
client (where the user is working directly with the data or uploading their own gene sets) and the backend (where data
is stored to allow it to persist between sessions) is consistent and clear.

## Product Requirements

### Users should be able to export differential expression results

Currently cellxgene users can choose two sets of cells and compute/display the log fold and adjusted p-value of the top ten differentially expressed genes. Additionally we display a histogram of the gene expression levels of the selected cells for each gene in the right side bar.

- [US1] A user should be able to export a gene set, if the gene set was created by differential expression, the pvalue, log fold change, and category label fields will be populated in addition to the gene name. This will be available in a csv file (see data formats for more detail).

> ### Users should be able to upload their own gene sets and see them in the right side bar with the option to apply them within current cellxgene functionalities, if the same gene ontology was used when generating the current dataset and the gene set uploaded by the user (ie the gene names match)

Reviewer: ```suggestion
### Users can upload their own gene sets and see them in the right side bar with the full range of current cellxgene functionalities.
```

Notes:
- Use "can" to be more assertive about what functionality is available instead "should be able to".
- There is no "option" to apply right? Gene sets will have full functionality in cellxgene explorer and in fact, if the gene set names don't match, you throw an error (later down in the requirements).
- Drop validity requirements for simplicity -- they are written down later as a sub-requirement anyways.

- [US2] A user should be able to upload a gene set via a csv file (see data formats for more detail) for a dataset.
- [US3] A user should be able to upload a gene set they previously downloaded from cellxgene
- [US4] When a user uploads a gene set, that list of genes should appear in the right side bar.
- [US5] If the user was logged in when they uploaded a gene set that gene set should persist if they logout and back in or reload the page
- [US6] If the gene names match those used in the dataset the user should be able to use general cellxgene functionality
  (color by genes, calculate differential expression levels, etc.) on those genes.
- [US7] If a gene in the user uploaded gene set does not exist in the dataset the user will see an error message specifying the gene that wasnt in the dataset, but the remaining genes will be displayed.

> ### If there is a standard format for differential expression output, we should follow that. If not, the format should be as straightforward as possible while serving these above cases

Reviewer: Would delete this -- it's an open question that you investigated, not a product requirements since you've already defined the gene sets format.

### Nonfunctional requirements

> - Users should be able to export precomputed differential expression results in real time

Reviewer: What does "real time" mean?

Author: @signechambers1 do you have an estimate for this? I think its somewhat related to @bmccandless's work on precompute

> - Users should be able to upload gene sets and have them immediately appear in the right side bar

Reviewer: Would delete this -- it's a functional requirement that's already written above.

**Figure 2** Cellxgene Diff Expression button pre cell selection

![Cellxgene Data Schema](imgs/diff_expression_button_post.png)

**Figure 3** Cellxgene Diff Expression button post cell selection

#### Additional Functionality

The gene sets created by a user will persist between sessions. When an authenticated user navigates to a particular dataset
the client will send a GET request to {dataset_name}/api/v0.01/gene_sets retrieving all of the user's gene sets for
that dataset.

In the future we will limit the user's ability to create anonymous groups of cells. They will need to use the annotation
feature to create a category and label the cell group in order to select it for differential expression analysis.

The user will be able to export a CSV file containing one or all of the gene sets they created by selecting cell sets and
calculating the 10 most differentially expressed genes along with their logfold and p-values between those sets.

The comments section of the csv file will specify the two sets of cells being compared to create that gene set.

The name of the exported file will be based on the user name and a timestamp rounded to the nearest second.

### Keeping Consensus

> This feature will require processing multiple updates to the same gene set, due to our hosting architecture this creates

Reviewer 2: What sort of race conditions are we trying to protect against.   Is it multiple POST/PUT calls from the client in flight simultaneously.  Or multiple clients acting at the same time?

Author: For now its just multiple POST/PUT calls from one client, but I did want to make it possible for multiple clients in the future to allow for shared genesets/multiple editors

Reviewer 2: By shared gene sets, do you mean shared by multiple users?   The current model is to associate a gene set with a single user.  Or do you mean that a user can simultaneously edit a geneset on multiple clients?

Author: multiple users. But it shouldnt require any changes, just making sure that we arent making it impossible in future iterations

server as part of the request body. The server will confirm that the counter sent from the client matches the counter
stored in the database for that gene set.

- If the values match, the server will update the gene set as specified in the PUT
  request and update the counter by incrementing it by 1. The response to the client will include the updated counter value
  to be used in future PUT requests for that gene set.
- If the counter stored in the database does not match the counter sent by the client the server will immediately send
  an error message to the client, specifying the name and uuid of the gene set that is out of sync. The client will
  then send over the entire gene set including all linked genes as a POST request. Because the gene set name is already
  taken for that user/dataset, the server will delete the currently stored gene set and any associated genes and recreate
  it based on the information included in the POST request. This will generate a new UUID for the gene set, and reset the
  consensus counter.
  In the case that the server times out or does not respond to a PUT request from the client, the client can retry or

>   just send a POST request to recreate the gene set.

Reviewer 2: From reading this description I'm not entirely sure what the purpose of the PUT API is, except as a performance optimization.   It looks like the client can use a POST instead.   Since using a PUT has more potential for errors, maybe we should only support the POST?

Author: The post is for creating a new gene set, the put is for updates

Reviewer 2: My point is that if the client encounters an error with PUT, then according to the above description, it will then use POST to re-create the gene set.    So, to simplify things, what if the client just started with POST?   In other words when is it beneficial ore preferable to use PUT instead of POST?   
It might be simpler to think of gene sets are immutable.  And they just get replaced if there are ever any changes.  If we do that we can get rid of PUT, and the need for a consensus counter, I think.

Author: That would definitely work, and probably be simpler. The thinking here is we will eventually use the same process of incremental updates for annotations as well. This is just an opportunity to try it out and work out any bugs on smaller/easier to debug sets of data.

  Because recreating the gene set assigns it a new UUID, any pending or high latency PUT requests sent prior to the update
  be ignored because they reference a UUID that no longer exists.

In the desktop version of the application the gene sets (and genes) will be stored locally as a csv under the name of
the dataset and the user session. On updates the application will create and store a new csv, maintaining the last
10 updates (this will match the current functionality for annotations in the desktop environment)

### Data Lifecycle

#### Creation

> Gene sets are created via a POST request to `/dataset_name/api/v0.1/gene_sets/`. This request is sent

Reviewer 2: gene_sets of var_sets.   We had a discussion about this at the last design meeting, not sure if we landed on a a decision.

Author: If we update the api to be var_sets do you think we should also update the table in the db?

Reviewer 2: Yes, that make sense.

- after a user uploads a csv continaing one (or more) gene sets.
- based on the results of comparing differential expression of two groups of cells.
- after the user creates a new gene set via the application GUI
  Genes are created as a part of the gene set creation step. They can also be added to an existing gene set via a PUT
  requst to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`

#### Updates

Gene sets are updated via a PUT request to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`. Genes can also be
updated via this mechanism

#### Deletion

Gene sets are deleted via a DELETE request to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`. All genes attached to
the gene set will be deleted as well.
Genes can be removed from a gene set by including their name in the `genes_to_delete` list that is part of the
request body for PUT requests to `/dataset_name/api/v0.1/gene_sets/{gene_set_uuid}`

### APIs

All APIs assume user identification is available with the request and the dataset being worked on is specified in the url

> Future API endpoints for consideration

Reviewer 2: Did you mean for these to be marked as future endpoints?   I think the one to get a list al all gene sets would be necessary.

Author: I didnt think they were necessary for the mvp, do you think they are?

Reviewer 2: Aren't these the endpoints that the client uses to get the current gene sets?

Author: no. The client can use GET dataset_name/api/v0.1/gene_sets/
That will return all of the genesets associated with the user/dataset

> - LIST dataset_name/api/v0.1/gene_sets/ - list all gene sets for a dataset (just names)

Reviewer 2: The comment at "just names" does not match the next section where we get the full information about the gene sets.

Author: the list endpoint would just return names, the get endpoint will return the full information.

Reviewer 2: Is GET just for a simple gene set, like the comment says "get all information for one particular gene set".
Or is it a list of gene sets, like the example output implies.

If it is for a single gene set, then where does the user get the gene_set_uuid from?

Author: GET dataset_name/api/v0.1/gene_sets/ will return all of the genesets and all of their genes associated with a user/dataset
I was thinking that LIST dataset_name/api/v0.1/gene_sets/ could be a future endpoint to just get the list of gene sets (name/uuid) for a dataset/user. And GET dataset_name/api/v0.1/gene_sets/{gene_set_uuid} would then  give you all of the genes/other information for a gene set

- GET dataset_name/api/v0.1/gene_sets/{gene_set_uuid} - get all information for one particular gene set

#### Get dataset_name/api/v0.1/gene_sets/

Return all of the gene sets (and accompanying genes) associated with a given user/dataset.

**Request:**
| Parameter | Description |
| ------------ | --------------------------------------------------------------------------------------------- |
| dataset_name | Identifies the dataset the gene sets are linked to. |
| user_id | Identifies the user linked to the gene set. |

**Response:**

| Code | Description                              |
| ---- | ---------------------------------------- |
| 200  | The gene sets were successfully returned |

| Key  | Description                                                                         |
| ---- | ----------------------------------------------------------------------------------- |
| data | List of gene set dicts containing their identifying information and a list of genes |

```json
{
  "gene_sets": [
    {
      "gene_set_name": "gene set name",
      "gene_set_uuid": "uuid",

>       "consensus_counter": "0001",

Reviewer 2: Can the consensus_counter be an integer instead?

| ---- | ------------------------------------------------------------ |
| 403  | User id invalid, or user does not have access to the dataset |

#### PUT dataset_name/api/v0.1/gene_sets/{gene_set_uuid}

Any changes to a gene set (the gene set itself or genes within that gene set) will be sent via json to the backend.
This includes the creation/deletion of genes within a gene set
Updates to comments will overwrite the currently stored comments
**Request:**

| Parameter     | Description                                                                                  |
| ------------- | -------------------------------------------------------------------------------------------- |
| dataset_name  | Identifies the dataset the gene sets are linked to.                                          |
| user_id       | Identifies the user creating the gene set                                                    |
| gene_set_uuid | UUID of gene set to be updated                                                               |
| body          | JSON formatted dict containing a gene set (and accompanying genes) to be updated-- see below |

> ```json

Reviewer 2: This interface adds some complexity to the client.  The client will need to have accurate knowledge of the state of the database, and the new state it wants for the gene set, in order to produce the add/delete/update lists.    The interface is not idempotent, so retrying after a failure may fail.   So if a failure does happen, the client has to do a POST.

Author: Yes. @mckinsel suggested having the client own the counter and when a request with the wrong counter number came in then the server could tell the client what the latest update it had processed was and the client could either resend the next the updates or send a post request recreating the gene set

Because gene sets are closely tied to differential expression, some aspects of the differential expression design plans will also be detailed below.

### User uploaded gene sets

#### 1. User creates csv containing a list of genes (and optional comments about why each gene was included)

CSV must follow format described in the Data Model section

#### 2. User uploads file to hosted cellxgene client

The user will typically upload a file from their local machine but we should consider supporting uploads from s3 in future iterations.

The user must be logged in to upload a gene set.

Users will upload the file by passing its path to the client.
The client will validate the format of the csv and check that the genes listed in the csv exist in the dataset

- if the format is invalid the client will raise a descriptive error message to the user
- if the file is valid, the client will display the genes in the right side bar
- if the file is valid, but it references genes that are not in the dataset, the client will display the genes that do
  exist in the right side bar and alert the user that their var set contained genes that are not a part of the dataset and they have

>   been removed from the var set that will be stored/displayed.

Reviewer 3: Will you make a distinction between genes that don't exist in the entire AnnData object and genes that are in the raw data, but missing from additional layers (normalized/subsetted versions) of the data?

In the future we will limit the user's ability to create anonymous groups of cells. They will need to use the annotation
feature to create a category and label the cell group in order to select it for differential expression analysis.

The user will be able to export a CSV file containing one or all of the gene sets they created by selecting cell sets and
calculating the 10 most differentially expressed genes along with their logfold and p-values between those sets.

The comments section of the csv file will specify the two sets of cells being compared to create that gene set.

The name of the exported file will be based on the user name and a timestamp rounded to the nearest second.

### Keeping Consensus

This feature will require processing multiple updates to the same gene set, due to our hosting architecture this creates
the potential for race conditions and concerns about keeping the data in sync between the database, server and client.  
To ensure data consistency, when the user is interacting with the data, the client is always the source of truth.
If what is stored in the database doesnt match what the user is seeing, the data in the database will be discarded and

> recreated based on the state of the client.

Reviewer 4: But in hosted, over time, there are potentially many clients looking at the same dataset.
